core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java

```diff
@@ -228,7 +228,7 @@ public InternalAggregation doReduce(List<InternalAggregation> aggregations, Redu
             }
         }
 
-        final int size = Math.min(requiredSize, buckets.size());
+        final int size = reduceContext.isFinalReduce() == false ? buckets.size() : Math.min(requiredSize, buckets.size());
```

Comment:
Since one of the goals of this change is to limit memory usage, I wonder whether this should use `getShardSize()` rather than `buckets.size()`, this should be a good trade-off between accuracy and memory usage? cc @colings86